### PR TITLE
[#294, #295] Added 'flare message' component, used for success or err…

### PR DESCRIPTION
**Pull request by millevlada**
…or notifications.

Added global error handler at UI that shows any error in client/server communication as error flare message.
Added support on server side to return custom error messages where appropriate.
Implemented success notification on several show case places at frontend.
Also added some code cleanup regarding response entity builder, where it was in focus ;)

### CODE REVIEW

File: web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java [hunk 1]
```diff
@@ -139,19 +138,17 @@ public ResponseEntity updateAnswers(@RequestBody SubmitTestRequest submitTestReq
...
     if (!tokenStore.isValidToken(token)) {
-      return HttpStatus.UNAUTHORIZED; // 401
+      throw new AccessDeniedException("Invalid applicant token!");
     }
 
     User user = tokenStore.getUser(token);
 
     if (!isAvailableForApplicant(user, test)) {
-      return HttpStatus.FORBIDDEN; // 403
+      throw new CodedException(HttpStatus.FORBIDDEN, "Test is not available for logged in applicant!");
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/TestController.java]: Same as above, please introduce a specific exception.

File: web/src/main/java/eu/execom/testgenerator/web/controller/UserController.java [hunk 2]
```diff
@@ -47,10 +48,10 @@
   public ResponseEntity getCurrentUser(@RequestHeader(name = AUTHENTICATION_TOKEN_KEY) String authToken) {
     User user = tokenStore.getUser(authToken);
     if (user == null) {
-      return new ResponseEntity(HttpStatus.BAD_REQUEST);
+      throw new CodedException(HttpStatus.BAD_REQUEST, "Unable to resolve user from token!");
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/UserController.java]: I think this should also be an authorization exception, with the proper message.

File: web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java [hunk 3]
```diff
@@ -5,17 +5,15 @@
...
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestMethod;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/controller/ReviewController.java]: Please do not use group imports. Configure your IDE not to group them.

File: web/src/main/java/eu/execom/testgenerator/web/util/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -22,61 +37,106 @@
...
 
   private void logError(HttpServletRequest request, Exception exception) {
 
-    log.error("Requested URL: {}. Exception raised: {}", request.getRequestURI(), exception);
+    log(Level.ERROR, request, exception);
   }
+
+  private void log(Level logLevel, HttpServletRequest request, Exception exception) {
+
+    String msg = "Requested URL: {}. Exception raised: {}";
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/util/GlobalExceptionHandler.java]: Please extract this to a constant.

File: web/src/main/java/eu/execom/testgenerator/web/util/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -22,61 +37,106 @@
...
+        log.info(msg, request.getRequestURI(), exception);
+        break;
+      case TRACE:
+        log.trace(msg, request.getRequestURI(), exception);
+        break;
+      case WARN:
+        log.warn(msg, request.getRequestURI(), exception);
+        break;
+      case ERROR:
+         default:
```
Comment (stanislavgrujic) [inline web/src/main/java/eu/execom/testgenerator/web/util/GlobalExceptionHandler.java]: Please fix spacing.

File: web/src/main/frontend/src/base/errorHandler.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import FlareStore from './stores/FlareStore';
+
+/**
+ * In case of error, marks 'apiCallError' flag on wrapped promise and re-throws it.
+ * @param req
+ */
+export const withApiCall = req => {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/errorHandler.js]: Rename to `catchApiCallError`.

File: web/src/main/frontend/src/base/errorHandler.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  return req.catch(err => {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/errorHandler.js]: No need for curly braces, can do implicit return.

File: web/src/main/frontend/src/base/errorHandler.js (hunk 6 above)
Comment (iscekic) [reply web/src/main/frontend/src/base/errorHandler.js]: No need for curly braces.

File: web/src/main/frontend/src/base/stores/FlareStore.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  @observable type;
+
+  @action.bound
+  info(flareMsg) {
+    return this.addFlare(TYPE.INFO, flareMsg);
+  }
+
+  @action.bound
+  error(e) {
+    if (e.message && e.response && e.response.body && e.response.body.error) {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/FlareStore.js]: This should be simplified with something like https://ramdajs.com/docs/#path

File: web/src/main/frontend/src/base/stores/FlareStore.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        `Error - ${e.message}: ${e.response.text}`
+      );
+    } else {
+      return this.addFlare(TYPE.ERROR, `Error: ${e}`);
+    }
+  }
+
+  @action.bound
+  addFlare(flareType, flareMsg) {
+    let maxId = 0;
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/FlareStore.js]: I kind of dislike this. Why not have `id` as a property of this class and simply increment it on each new flare?

File: web/src/main/frontend/src/base/stores/FlareStore.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    if (this.flares.length > 0) {
+      maxId = this.flares[this.flares.length - 1].key;
+    }
+    this.flares.push({key: maxId + 1, type: flareType, msg: flareMsg});
+    this.type = flareType === TYPE.ERROR ? TYPE.ERROR : this.getType();
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/FlareStore.js]: Shouldn't type be a property of each individual flare?

File: web/src/main/frontend/src/base/errorHandler.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Default error handler.
+ * By default, all errors derived from API calls (marked with 'apiCallError') are handled
+ * as flare error notifications.
+ * Other rejected promises are re-thrown.
+ *
+ * @param err Rejected promise
+ */
+export const handleError = err => {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/errorHandler.js]: Rename to `handleApiError`.

File: web/src/main/frontend/src/base/errorHandler.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  });
+};
+
+/**
+ * Default error handler.
+ * By default, all errors derived from API calls (marked with 'apiCallError') are handled
+ * as flare error notifications.
+ * Other rejected promises are re-thrown.
+ *
+ * @param err Rejected promise
```
Comment (iscekic) [inline web/src/main/frontend/src/base/errorHandler.js]: err is an exception object, not a promise

File: web/src/main/frontend/src/components/ReviewTest.js [hunk 13]
```diff
@@ -4,15 +4,20 @@ import {inject, observer} from 'mobx-react';
...
   submit = () => {
-    this.props.completedTest.applyReview().then(() => {
-      this.props.completedTest.exitReview();
-      this.props.dashboard.fetchDashboardData();
-    });
+    this.props.completedTest
+      .applyReview()
+      .then(() => {
+        this.props.completedTest.exitReview();
+        this.props.dashboard.fetchDashboardData();
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ReviewTest.js]: If this call (or the one above) is async, please make sure it's intended that they happen in parallel.

File: web/src/main/frontend/src/components/ui/MsgFlare.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+ * Info message disappears automatically after 5 seconds on first mouse movement.
+ * Error message disappears automatically after user hover over message, after 5 seconds.
+ */
+@inject('flare')
+@observer
+class MsgFlare extends React.Component {
+  autoRemoveTimer;
+
+  constructor(props) {
+    super(props);
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ui/MsgFlare.js]: Unneeded constructor, you can directly set initial values on the class:

```js
class Whatever {
  state = {removing: false};

  autoremove = () => { ... };
}
```

File: web/src/main/frontend/src/components/ui/MsgFlare.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    window.addEventListener('mousemove', this.fadeOutMsg);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('mousemove', this.fadeOutMsg);
+    this.autoRemoveTimer = null;
+  }
+
+  renderIcon = () => {
+    const icon =
```
Comment (iscekic) [inline web/src/main/frontend/src/components/ui/MsgFlare.js]: Please use `if` instead of nested ternaries.

**Pull request merged**